src/task/callback.rs

```diff
@@ -0,0 +1,265 @@
+}
+
+impl<Spawn> Task<Spawn>
+where
+    Spawn: LocalSpawn,
+{
+    /// Creates a [`FnOnce`] task.
+    pub fn new_once(t: impl FnOnce(&mut Handle<'_, Spawn>) + Send + 'static) -> Self {
+        Task::Once(Box::new(t))
+    }
+
+    /// Creates a [`FnMut`] task.
+    pub fn new_mut(t: impl FnMut(&mut Handle<'_, Spawn>) + Send + 'static) -> Self {
+        Task::Mut(Box::new(t))
+    }
+}
+
+/// Handle passed to the task closure.
+///
+/// It can be used to spawn new tasks or control whether this task should be
+/// rerun.
+pub struct Handle<'a, Spawn>
+where
+    Spawn: LocalSpawn,
+{
+    spawn: &'a mut Spawn,
+    ctx: &'a Spawn::TaskContext,
+    rerun: bool,
+}
+
+impl<'a, Spawn> Handle<'a, Spawn>
+where
+    Spawn: LocalSpawn<Task = Task<Spawn>>,
+{
+    /// Spawns a [`FnOnce`] to the thread pool with the same task context.
+    pub fn spawn_once(&mut self, t: impl FnOnce(&mut Handle<'_, Spawn>) + Send + 'static) {
+        self.spawn.spawn_ctx(Task::new_once(t), self.ctx);
+    }
+
+    /// Spawns a [`FnMut`] to the thread pool with the same task context.
+    pub fn spawn_mut(&mut self, t: impl FnMut(&mut Handle<'_, Spawn>) + Send + 'static) {
+        self.spawn.spawn_ctx(Task::new_mut(t), self.ctx);
+    }
+
+    /// Spawns a [`FnOnce`] to the thread pool with the given task context.
+    pub fn spawn_once_ctx(
+        &mut self,
+        t: impl FnOnce(&mut Handle<'_, Spawn>) + Send + 'static,
+        ctx: &Spawn::TaskContext,
+    ) {
+        self.spawn.spawn_ctx(Task::new_once(t), ctx);
+    }
+
+    /// Spawns a [`FnMut`] to the thread pool with the given task context.
+    pub fn spawn_mut_ctx(
+        &mut self,
+        t: impl FnMut(&mut Handle<'_, Spawn>) + Send + 'static,
+        ctx: &Spawn::TaskContext,
+    ) {
+        self.spawn.spawn_ctx(Task::new_mut(t), ctx);
+    }
+
+    /// Gets the task context.
+    pub fn context(&self) -> &Spawn::TaskContext {
+        &self.ctx
+    }
+
+    /// Sets whether this task should be rerun later.
+    pub fn set_rerun(&mut self, rerun: bool) {
+        self.rerun = rerun;
+    }
+}
+
+/// Callback task runner.
+///
+/// It's possible that a task can't be finished in a single execution and needs
+/// to be rerun. `max_inspace_spin` is the maximum times a task is rerun at once
+/// before being put back to the thread pool.
+pub struct Runner<Spawn> {
+    max_inplace_spin: usize,
+    _phantom: PhantomData<Spawn>,
+}
+
+impl<Spawn> Runner<Spawn> {
+    /// Creates a new runner with given `max_inplace_spin`.
+    pub fn new(max_inplace_spin: usize) -> Self {
+        Self {
+            max_inplace_spin,
+            ..Default::default()
+        }
+    }
+
+    /// Sets `max_inplace_spin`.
+    pub fn set_max_inplace_spin(&mut self, max_inplace_spin: usize) {
+        self.max_inplace_spin = max_inplace_spin;
+    }
+}
+
+impl<Spawn> Default for Runner<Spawn> {
+    fn default() -> Self {
+        Runner {
+            max_inplace_spin: 3,
```

Comment:
Why use 3 instead of 4? How does it perform?

Reply:
I just keep the behavior the same as the original code. I guess the default value is by rule of thumb and highly workload related (for callback tasks). Maybe it's not that important.

Reply:
But it's 4 in the original code: https://github.com/BusyJay/chocolates/blob/dev/src/thread_pool/callback.rs#L89

Reply:
4 in that code is the same as 3 here because the value in that code is the maximum run times and here is rerun times.